src/mame/drivers/att4425.cpp

```diff
@@ -0,0 +1,300 @@
+    Also known as Teletype model 56D.
+
+    Skeleton driver.  This terminal is supported by netpbm -- pbmto4425(1)
+
+    To do:
+    - keyboard
+    - everything else
+
+****************************************************************************/
+
+
+#include "emu.h"
+
+#include "cpu/z80/z80.h"
+#include "machine/clock.h"
+#include "machine/i8251.h"
+#include "machine/ram.h"
+#include "machine/z80ctc.h"
+#include "machine/z80sio.h"
+
+#include "bus/rs232/rs232.h"
+#include "machine/keyboard.h"
+
+#include "screen.h"
+
+
+#define SCREEN_TAG          "screen"
+#define Z80_TAG             "maincpu"
+#define Z80CTC_TAG          "z80ctc"
+#define Z80SIO_TAG          "z80sio"
+#define RS232_A_TAG         "sioa"
+#define RS232_B_TAG         "siob"
+#define I8251_TAG           "i8251"
+
+class att4425_state : public driver_device
+{
+public:
+	att4425_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, Z80_TAG)
+		, m_i8251(*this, I8251_TAG)
+		, m_sio(*this, Z80SIO_TAG)
+		, m_p_videoram(*this, "videoram")
+		, m_p_chargen(*this, "chargen")
+		, m_screen(*this, "screen")
+		{ }
+
+	DECLARE_PALETTE_INIT(att4425);
+
+	DECLARE_WRITE_LINE_MEMBER(write_line_clock);
+	DECLARE_WRITE_LINE_MEMBER(write_keyboard_clock);
+
+	virtual void machine_start() override;
+
+	virtual void video_start() override;
+	uint32_t screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+
+	required_device<z80_device> m_maincpu;
+	required_device<i8251_device> m_i8251;
+	required_device<z80sio_device> m_sio;
+	required_shared_ptr<uint8_t> m_p_videoram;
+	required_region_ptr<u8> m_p_chargen;
+	required_device<screen_device> m_screen;
+};
+
+/* Memory Maps */
+
+static ADDRESS_MAP_START( att4425_mem, AS_PROGRAM, 8, att4425_state )
+	AM_RANGE(0x0000, 0x7fff) AM_ROM AM_REGION(Z80_TAG, 0)
+	AM_RANGE(0x8000, 0xffff) AM_RAM AM_SHARE("videoram") // c000..f7af?
+ADDRESS_MAP_END
+
+static ADDRESS_MAP_START( att4425_io, AS_IO, 8, att4425_state )
+	AM_RANGE(0x0000, 0x0000) AM_MIRROR(0xff00) AM_DEVREADWRITE(I8251_TAG, i8251_device, data_r, data_w)
+	AM_RANGE(0x0001, 0x0001) AM_MIRROR(0xff00) AM_DEVREADWRITE(I8251_TAG, i8251_device, status_r, control_w)
+	AM_RANGE(0x0008, 0x0008) AM_MIRROR(0xff00) AM_UNMAP // write 08
+	AM_RANGE(0x0010, 0x0010) AM_MIRROR(0xff00) AM_UNMAP // write 10
+	AM_RANGE(0x0014, 0x0015) AM_MIRROR(0xff00) AM_UNMAP // NVRAM? read 14, 15; write 14
+	AM_RANGE(0x0018, 0x001b) AM_MIRROR(0xff00) AM_DEVREADWRITE(Z80CTC_TAG, z80ctc_device, read, write)
+	AM_RANGE(0x001c, 0x001f) AM_MIRROR(0xff00) AM_DEVREADWRITE(Z80SIO_TAG, z80sio_device, ba_cd_r, ba_cd_w)
+ADDRESS_MAP_END
+
+/* Input Ports */
+
+static INPUT_PORTS_START( att4425 )
+INPUT_PORTS_END
+
+/* Video */
+
+PALETTE_INIT_MEMBER(att4425_state, att4425)
+{
+	palette.set_pen_color(0, 0, 0, 0);   /* Black */
+	palette.set_pen_color(1, 0, 255, 0); /* Normal */
+	palette.set_pen_color(2, 0, 192, 0); /* Dim */
```

Comment:
We already have a setup for doing monochrome with hilight, see e.g. osborne1.cpp for a example.  The phosphor colour is an attribute of the screen, and a monochrome palette is used.

```
    MCFG_SCREEN_ADD_MONOCHROME("screen", RASTER, rgb_t::green())
    MCFG_SCREEN_UPDATE_DRIVER(osborne1_state, screen_update)
    MCFG_SCREEN_RAW_PARAMS( MAIN_CLOCK, 1024, 0, 104*8, 260, 0, 24*10 )
    MCFG_SCREEN_PALETTE("palette")
    MCFG_GFXDECODE_ADD("gfxdecode", "palette", osborne1)
    MCFG_PALETTE_ADD_MONOCHROME_HIGHLIGHT("palette")
```